### ISSUE TITLE
cf push fails to upload app when app package is already in a failed state

### ISSUE THREAD

**Issue by thausler786** [cloudfoundry/cli#1216]
## Command  
`cf push usage-service-scheduler`

## What occurred
Very first push: We pushed our app and it failed because the package was too large (we had log files that had grown out of control locally)
```
cf push usage-service-scheduler
Using manifest file a1-scheduler.yml

Using stack cflinuxfs2...
OK
Updating app usage-service-scheduler in org pivotal / space smoketest as cf-fe-eng+a1@pivotallabs.com...
OK

App usage-service-scheduler is a worker, skipping route creation
Uploading usage-service-scheduler...
Uploading app files from: /Users/pivotal/workspace/app-usage-service
Uploading 530M, 410 files
Done uploading
FAILED
Error processing app files: Error uploading application.
The app package is invalid: Package may not be larger than 1073741824 bytescf push usage-service-scheduler -f a1-scheduler.yml
Using manifest file a1-scheduler.yml

Using stack cflinuxfs2...
OK
Updating app usage-service-scheduler in org pivotal / space smoketest as cf-fe-eng+a1@pivotallabs.com...
OK

App usage-service-scheduler is a worker, skipping route creation
Uploading usage-service-scheduler...
Uploading app files from: /Users/pivotal/workspace/app-usage-service
Uploading 530M, 410 files
Done uploading
FAILED
Error processing app files: Error uploading application.
The app package is invalid: Package may not be larger than 1073741824 bytes
```
Subsequent pushes: We removed the troublesome log files and attempted to re-push. These failed with
```
cf push usage-service-scheduler -f a1-scheduler.yml
Using manifest file a1-scheduler.yml

Using stack cflinuxfs2...
OK
Updating app usage-service-scheduler in org pivotal / space smoketest as cf-fe-eng+a1@pivotallabs.com...
FAILED
Server error, status code: 400, error code: 150001, message: The app package is invalid: bits have not been uploaded
```

## What you expected to occur
We expect the CLI to re-upload the application, no amount of change in the code locally forced an app upload.

## CLI Version 
`cf version 6.29.2+c66d0f342.2017-08-25`

## CC API Endpoint Version  

2.94.0

## CF Trace

```
 2017-08-30 14:16:45 ⌚ ruby 2.3.1p112 bartol in ~/workspace/app-usage-service
± dr+th |master {1} U:1 ?:2 ✗| → cf push -v usage-service-scheduler -f a1-scheduler.yml
Using manifest file a1-scheduler.yml


REQUEST: [2017-08-30T14:19:25-07:00]
POST /oauth/token HTTP/1.1
Host: login.a1.cf-app.com
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Connection: close
Content-Type: application/x-www-form-urlencoded
User-Agent: go-cli 6.29.2+c66d0f342.2017-08-25 / darwin


RESPONSE: [2017-08-30T14:19:25-07:00]
HTTP/1.1 200 OK
Connection: close
Content-Length: 2440
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Type: application/json;charset=UTF-8
Date: Wed, 30 Aug 2017 21:19:25 GMT
Expires: 0
Pragma: no-cache
Strict-Transport-Security: max-age=31536000 ; includeSubDomains
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Vcap-Request-Id: 5fbebb28-415b-4ff3-71be-c3abd7e925af
X-Xss-Protection: 1; mode=block

{"access_token":"[PRIVATE DATA HIDDEN]","token_type":"[PRIVATE DATA HIDDEN]","refresh_token":"[PRIVATE DATA HIDDEN]","expires_in":599,"scope":"cloud_controller.read password.write cloud_controller.write openid uaa.user","jti":"d61e642fc0644ed7876b496d05434398"}
Using stack cflinuxfs2...

REQUEST: [2017-08-30T14:19:25-07:00]
GET /v2/stacks?q=name%3Acflinuxfs2 HTTP/1.1
Host: api.a1.cf-app.com
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.29.2+c66d0f342.2017-08-25 / darwin


RESPONSE: [2017-08-30T14:19:26-07:00]
HTTP/1.1 200 OK
Connection: close
Content-Length: 460
Content-Type: application/json;charset=utf-8
Date: Wed, 30 Aug 2017 21:19:26 GMT
Server: nginx
X-Content-Type-Options: nosniff
X-Vcap-Request-Id: 892f944f-83f4-4b0e-5008-6208c5e3ab1d
X-Vcap-Request-Id: b9c8805e-9f2e-4cb8-6b1f-3434769ab3b3::e07dbf8a-ab04-44ed-9b15-5fd832d5fd14

{
  "total_results": 1,
  "total_pages": 1,
  "prev_url": null,
  "next_url": null,
  "resources": [
    {
      "metadata": {
        "guid": "45494bd4-b9d8-4c11-97f2-7ad724b8208f",
        "url": "/v2/stacks/45494bd4-b9d8-4c11-97f2-7ad724b8208f",
        "created_at": "2015-02-24T19:38:39Z",
        "updated_at": null
      },
      "entity": {
        "name": "cflinuxfs2",
        "description": "Cloud Foundry Linux-based filesystem"
      }
    }
  ]
}
OK

REQUEST: [2017-08-30T14:19:26-07:00]
GET /v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/apps?q=name%3Ausage-service-scheduler&inline-relations-depth=1 HTTP/1.1
Host: api.a1.cf-app.com
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.29.2+c66d0f342.2017-08-25 / darwin



RESPONSE: [2017-08-30T14:19:26-07:00]
HTTP/1.1 200 OK
Connection: close
Content-Length: 8530
Content-Type: application/json;charset=utf-8
Date: Wed, 30 Aug 2017 21:19:26 GMT
Server: nginx
X-Content-Type-Options: nosniff
X-Vcap-Request-Id: 3ecec817-cf2c-44a4-4a25-18010c65370a
X-Vcap-Request-Id: fd027362-14c2-4cbf-5928-702660c3ee17::918b7e47-2203-4a41-a929-d5ecf584798a

{
  "total_results": 1,
  "total_pages": 1,
  "prev_url": null,
  "next_url": null,
  "resources": [
    {
      "metadata": {
        "guid": "fa18ff8a-9c9b-4955-aff0-c37e031951ba",
        "url": "/v2/apps/fa18ff8a-9c9b-4955-aff0-c37e031951ba",
        "created_at": "2016-03-21T18:38:04Z",
        "updated_at": "2017-08-30T20:52:05Z"
      },
      "entity": {
        "name": "usage-service-scheduler",
        "production": false,
        "space_guid": "e6445eb3-fdac-4049-bafc-94d1703d5e78",
        "stack_guid": "45494bd4-b9d8-4c11-97f2-7ad724b8208f",
        "buildpack": "ruby_buildpack",
        "detected_buildpack": null,
        "detected_buildpack_guid": null,
        "environment_json": {
          "AIRBRAKE_API_KEY": "scrubbed",
          "AIRBRAKE_ENV": "scrubbed",
          "ALLOW_TEST_DATA_CREATION": "false",
          "ALLOW_VIEWING_LOGS": "false",
          "BUNDLE_WITHOUT": "test:development",
          "CF_CC_API_URL": "https://api.a1.cf-app.com",
          "CF_CLIENT_ID": "scrubbed",
          "CF_CLIENT_SECRET": "scrubbed",
          "CF_UAA_SERVER_URL": "https://uaa.a1.cf-app.com",
          "CLOUD_CONTROLLER_URL": "https://api.a1.cf-app.com",
          "DATABASE_URL": "scrubbed",
          "DELAYED_JOB_UI_PASSWORD":"[PRIVATE DATA HIDDEN]",
          "DELAYED_JOB_UI_USERNAME": "usage",
          "ENABLE_DELAYED_JOB_UI": "true",
          "ENABLE_SYSTEM_SERVICE_USAGE_REPORT": "false",
          "EVENT_FETCHING_OFFSET": "60",
          "NEW_RELIC_AGENT_ENABLED": "true",
          "NEW_RELIC_APP_NAME": "Usage Service (Smoketest)",
          "NEW_RELIC_LICENSE_KEY": "scrubbed",
          "NEW_RELIC_MONITOR_MODE": "true",
          "RACK_ENV": "production",
          "RAILS_ENV": "production",
          "SECRET_KEY_BASE": "scrubbed",
          "SKIP_OLD_MONTHLY_APP_USAGE_CONVERSION": "true",
          "SKIP_USAGE_CLEANUP": "true",
          "UAA_URL": "https://uaa.a1.cf-app.com",
          "USAGE_SERVICE_UAA_CLIENT_ID": "scrubbed",
          "USAGE_SERVICE_UAA_CLIENT_SECRET": "scrubbed",
          "VERIFY_SSL": "true"
        },
        "memory": 1024,
        "instances": 1,
        "disk_quota": 1024,
        "state": "STARTED",
        "version": "a314a922-97f7-4e0c-8882-5d6f031d3438",
        "command": "bundle exec rake scheduler:start",
        "console": false,
        "debug": null,
        "staging_task_id": "dfa27cb2-a83c-4452-ad09-f6e0fca6b44a",
        "package_state": "FAILED",
        "health_check_type": "none",
        "health_check_timeout": null,
        "health_check_http_endpoint": null,
        "staging_failed_reason": null,
        "staging_failed_description": null,
        "diego": true,
        "docker_image": null,
        "docker_credentials": {
          "username": null,
          "password": null
        },
        "package_updated_at": "2017-08-30T20:57:09Z",
        "detected_start_command": "",
        "enable_ssh": true,
        "ports": [
          8080
        ],
        "space_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78",
        "space": {
          "metadata": {
            "guid": "e6445eb3-fdac-4049-bafc-94d1703d5e78",
            "url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78",
            "created_at": "2013-05-16T22:08:04Z",
            "updated_at": "2013-06-20T19:21:34Z"
          },
          "entity": {
            "name": "smoketest",
            "organization_guid": "8f88362f-547c-4158-808b-4605468387d5",
            "space_quota_definition_guid": null,
            "isolation_segment_guid": null,
            "allow_ssh": true,
            "organization_url": "/v2/organizations/8f88362f-547c-4158-808b-4605468387d5",
            "developers_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/developers",
            "managers_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/managers",
            "auditors_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/auditors",
            "apps_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/apps",
            "routes_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/routes",
            "domains_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/domains",
            "service_instances_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/service_instances",
            "app_events_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/app_events",
            "events_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/events",
            "security_groups_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/security_groups",
            "staging_security_groups_url": "/v2/spaces/e6445eb3-fdac-4049-bafc-94d1703d5e78/staging_security_groups"
          }
        },
        "stack_url": "/v2/stacks/45494bd4-b9d8-4c11-97f2-7ad724b8208f",
        "stack": {
          "metadata": {
            "guid": "45494bd4-b9d8-4c11-97f2-7ad724b8208f",
            "url": "/v2/stacks/45494bd4-b9d8-4c11-97f2-7ad724b8208f",
            "created_at": "2015-02-24T19:38:39Z",
            "updated_at": null
          },
          "entity": {
            "name": "cflinuxfs2",
            "description": "Cloud Foundry Linux-based filesystem"
          }
        },
        "routes_url": "/v2/apps/fa18ff8a-9c9b-4955-aff0-c37e031951ba/routes",
        "routes": [

        ],
        "events_url": "/v2/apps/fa18ff8a-9c9b-4955-aff0-c37e031951ba/events",
        "service_bindings_url": "/v2/apps/fa18ff8a-9c9b-4955-aff0-c37e031951ba/service_bindings",
        "service_bindings": [
          {
            "metadata": {
              "guid": "76cee2fa-42c9-42c1-ad84-7610794bfce8",
              "url": "/v2/service_bindings/76cee2fa-42c9-42c1-ad84-7610794bfce8",
              "created_at": "2016-03-21T18:52:40Z",
              "updated_at": null
            },
            "entity": {
              "app_guid": "fa18ff8a-9c9b-4955-aff0-c37e031951ba",
              "service_instance_guid": "1f0258d8-ae4e-4449-9db9-7c41c4f86476",
              "credentials": {

              },
              "binding_options": {

              },
              "gateway_data": null,
              "gateway_name": "",
              "syslog_drain_url": "scrubbed",
              "volume_mounts": [

              ],
              "app_url": "/v2/apps/fa18ff8a-9c9b-4955-aff0-c37e031951ba",
              "service_instance_url": "/v2/user_provided_service_instances/1f0258d8-ae4e-4449-9db9-7c41c4f86476"
            }
          },
          {
            "metadata": {
              "guid": "e1b32a10-5878-4708-8ed4-2d7817320200",
              "url": "/v2/service_bindings/e1b32a10-5878-4708-8ed4-2d7817320200",
              "created_at": "2017-07-21T17:30:36Z",
              "updated_at": "2017-07-21T17:30:36Z"
            },
            "entity": {
              "app_guid": "fa18ff8a-9c9b-4955-aff0-c37e031951ba",
              "service_instance_guid": "f32eaf5c-db66-4689-a6fb-6378915b15f6",
              "credentials": {
                "hostname": "scrubbed",
                "port": 3306,
                "name": "cf_f32eaf5c_db66_4689_a6fb_6378915b15f6",
                "username": "scrubbed",
                "password":"[PRIVATE DATA HIDDEN]",
                "uri": "scrubbed",
                "jdbcUrl": "scrubbed"
              },
              "binding_options": {

              },
              "gateway_data": null,
              "gateway_name": "",
              "syslog_drain_url": null,
              "volume_mounts": [

              ],
              "app_url": "/v2/apps/fa18ff8a-9c9b-4955-aff0-c37e031951ba",
              "service_instance_url": "/v2/service_instances/f32eaf5c-db66-4689-a6fb-6378915b15f6"
            }
          }
        ],
        "route_mappings_url": "/v2/apps/fa18ff8a-9c9b-4955-aff0-c37e031951ba/route_mappings"
      }
    }
  ]
}
Updating app usage-service-scheduler in org pivotal / space smoketest as cf-fe-eng+a1@pivotallabs.com...

REQUEST: [2017-08-30T14:19:26-07:00]
PUT /v2/apps/fa18ff8a-9c9b-4955-aff0-c37e031951ba?async=true&inline-relations-depth=1 HTTP/1.1
Host: api.a1.cf-app.com
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.29.2+c66d0f342.2017-08-25 / darwin

{"name":"usage-service-scheduler","command":"bundle exec rake scheduler:start","instances":1,"memory":1024,"disk_quota":1024,"stack_guid":"45494bd4-b9d8-4c11-97f2-7ad724b8208f","buildpack":"ruby_buildpack","environment_json":{"AIRBRAKE_API_KEY":"scrubbed","AIRBRAKE_ENV":"Smoketest","ALLOW_TEST_DATA_CREATION":"false","ALLOW_VIEWING_LOGS":"false","BUNDLE_WITHOUT":"test:development","CF_CC_API_URL":"https://api.a1.cf-app.com","CF_CLIENT_ID":"scrubbed","CF_CLIENT_SECRET":"scrubbed","CF_UAA_SERVER_URL":"https://uaa.a1.cf-app.com","CLOUD_CONTROLLER_URL":"https://api.a1.cf-app.com","DATABASE_URL":"scrubbed","DELAYED_JOB_UI_PASSWORD":"[PRIVATE DATA HIDDEN]","DELAYED_JOB_UI_USERNAME":"scrubbed","ENABLE_DELAYED_JOB_UI":"true","ENABLE_SYSTEM_SERVICE_USAGE_REPORT":"false","EVENT_FETCHING_OFFSET":"60","NEW_RELIC_AGENT_ENABLED":"true","NEW_RELIC_APP_NAME":"Usage Service (Smoketest)","NEW_RELIC_LICENSE_KEY":"scrubbed","NEW_RELIC_MONITOR_MODE":"true","RACK_ENV":"production","RAILS_ENV":"production","SECRET_KEY_BASE":"scrubbed","SKIP_OLD_MONTHLY_APP_USAGE_CONVERSION":"true","SKIP_USAGE_CLEANUP":"true","UAA_URL":"https://uaa.a1.cf-app.com","USAGE_SERVICE_UAA_CLIENT_ID":"scrubbed","USAGE_SERVICE_UAA_CLIENT_SECRET":"scrubbed","VERIFY_SSL":"true"},"health_check_type":"none","ports":[8080]}

RESPONSE: [2017-08-30T14:19:27-07:00]
HTTP/1.1 400 Bad Request
Connection: close
Content-Length: 137
Content-Type: application/json;charset=utf-8
Date: Wed, 30 Aug 2017 21:19:27 GMT
Server: nginx
X-Content-Type-Options: nosniff
X-Vcap-Request-Id: 09a3cd23-2f7f-443b-63d9-e18b9d658c01
X-Vcap-Request-Id: 7d04264c-f7e2-4f8d-6f8c-64d18454927c::ace38553-9fe2-496a-9791-09078c77790c

{
  "description": "The app package is invalid: bits have not been uploaded",
  "error_code": "CF-AppPackageInvalid",
  "code": 150001
}

FAILED
Server error, status code: 400, error code: 150001, message: The app package is invalid: bits have not been uploaded
```

## Platform & Shell Details  

Mac OS X 10.11 iTerm

**Comment by dkoper** [comment]
@zrob It seems CC is not allowing us to make an update to a (v2) app before we upload app bits. Any particular reason for that?

**Comment by zrob** [comment]
I can't think of a specific reason, the only update that I think should be invalid would be mark the app as started with an invalid package.

**Comment by dkoper** [comment]
Thanks @zrob, great hint.

I was able to reproduce this:
1. First I push an app and start it successfully.
1. When I then update the app with a big package with `cf push bigapp -p bigfile.zip`, I get the `Package may not be larger than 1073741824 bytes` error.
1. When I then push a smaller file with `cf push bigapp -p smallfile.zip`, it works fine. Trace includes:
```
REQUEST: [2017-09-18T14:20:02+10:00]
PUT /v2/apps/411589dc-6592-4428-94af-5ff82ec61123?async=true&inline-relations-depth=1 HTTP/1.1
...
{"name":"bigapp"}
```

However, when I use an app manifest with other entries (instances, memory, stack, disk_quota) and follow the steps above, the API call above becomes:
```
PUT /v2/apps/411589dc-6592-4428-94af-5ff82ec61123?async=true&inline-relations-depth=1 HTTP/1.1
...
{"name":"bigapp","instances":1,"memory":32,"disk_quota":1024,"stack_guid":"0fd54a39-4925-4652-8036-81f8300fd01c","environment_json
":{}}
```

@zrob Even though none of those additional attributes have changed from the previous push, could one of those be triggering the error?

---

I tried it with refactored push as well. Although the API request is a bit different, the error is the same:
```
PUT /v2/apps/411589dc-6592-4428-94af-5ff82ec61123 HTTP/1.1
...
{
  "disk_quota": 1024,
  "health_check_type": "port",
  "instances": 1,
  "memory": 32,
  "name": "bigapp",
  "state": "STARTED"
}
```

This one matches Zach's comment (marking the app as started with an invalid package).

Finally, I tried with V3 app push to do the update above, and that completed successfully. But this command doesn't support the app manifest, nor any flags to set the additional flags (memory, etc.) yet, so it didn't make the equivalent API call.
